// @angular-builders/jest/src/default-config.resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@angular-devkit/core");
const default_config_1 = require("./jest-config/default-config");
exports.testPattern = `/**/+(*.)+(spec|test).+(ts|js)?(x)`;
exports.tsConfigName = 'tsconfig.spec.json';
exports.jestPresetRootPath = `node_modules/jest-preset-angular`;
exports.preprocessor = 'preprocessor.js';
exports.snapshotSerializer = 'AngularSnapshotSerializer.js';
exports.htmlCommentSerializer = 'HTMLCommentSerializer.js';
exports.transformRegex = '^.+\\.(ts|js|html)$';
class DefaultConfigResolver {
    resolveGlobal(workspaceRoot) {
        const jestPresetFullPath = core_1.join(workspaceRoot, exports.jestPresetRootPath);
        return Object.assign({}, default_config_1.default, { transform: {
                [exports.transformRegex]: core_1.getSystemPath(core_1.join(jestPresetFullPath, exports.preprocessor))
            }, snapshotSerializers: [
                core_1.getSystemPath(core_1.join(jestPresetFullPath, exports.snapshotSerializer)),
                core_1.getSystemPath(core_1.join(jestPresetFullPath, exports.htmlCommentSerializer))
            ] });
    }
    resolveForProject(projectRoot) {
        return {
            globals: {
                'ts-jest': {
                    tsConfigFile: core_1.getSystemPath(core_1.join(projectRoot, exports.tsConfigName))
                }
            },
            testMatch: [
                `${core_1.getSystemPath(core_1.join(projectRoot, exports.testPattern))}`
            ]
        };
    }
}
exports.DefaultConfigResolver = DefaultConfigResolver;
